site/tutorials/http-api-authorization-tokens/index.md

```diff
@@ -0,0 +1,269 @@
+---
+
+{% contentfor header %}
+# HTTP API Authorization Using Tokens
+
+Anything that exposes an HTTP API (whether an individual microservice or an application as a whole) needs to control who can run those APIs and when.  OPA makes it easy to write fine-grained, context-aware policies to implement API authorization. It also supports sending information useful to policy decisions using JSON Web Tokens.
+
+{% endcontentfor %}
+
+{% contentfor body %}
+
+## Goals
+
+In this tutorial, you'll use a simple HTTP web server that accepts any HTTP GET
+request that you issue and echoes the OPA decision back as text. Both OPA and
+the web server will be run as containers.
+
+For this tutorial, our desired policy is:
+
+* People can see their own salaries (`GET /finance/salary/{user}` is permitted for `{user}`)
+* People can see the salaries of their direct subordinates (`GET /finance/salary/{user}` is permitted for `{user}`'s manager)
+
+## Prerequisites
+
+This tutorial requires [Docker Compose](https://docs.docker.com/compose/install/) to run a demo web server along with OPA.
+
+## Steps
+
+### 1. Bootstrap the tutorial environment using Docker Compose.
+
+First, create a docker-compose.yml file that runs OPA and the demo web server.
+
+```shell
+cat >docker-compose.yml <<EOF
+version: '2'
+services:
+  opa:
+    image: openpolicyagent/opa:0.4.9
+    ports:
+      - 8181:8181
+    command:
+      - "run"
+      - "--server"
+      - "--log-level=debug"
+  api_server:
+    image: openpolicyagent/demo-restful-api-tokens:latest
+    ports:
+      - 5000:5000
+    environment:
+      - OPA_ADDR=http://opa:8181
+      - POLICY_PATH=/v1/data/httpapi/authz
+EOF
+```
+{: .opa-collapse--ignore}
+
+Then run `docker-compose` to pull and run the containers.
+
+```shell
+docker-compose -f docker-compose.yml up
+```
+
+### 2. Load a simple policy into OPA.
+
+In another terminal, create a simple policy. The policy below allows users to
+request their own salary as well as the salary of their direct subordinates.
+
+```shell
+cat >example.rego <<EOF
+package httpapi.authz
+
+import input as http_api
+
+p = [header, payload, sign] { io.jwt.decode(http_api.token, header, payload, sign) }
+user_owns_token { http_api.user = p[1].name }
+
+default allow = false
+allow {
+  http_api.method = "GET"
+  http_api.path = ["finance", "salary", username]
+  username = p[1].name
+  user_owns_token
+}
+
+allow {
+  http_api.method = "GET"
+  http_api.path = ["finance", "salary", username]
+  p[1].subordinates[_] = username
+  user_owns_token
+}
+EOF
+```
+{: .opa-collapse--ignore}
+
+Then load the policy via OPA's REST API.
+
+```shell
+curl -X PUT --data-binary @example.rego \
+  localhost:8181/v1/policies/example
+```
+
+### 3. Check that `alice` can see her own salary.
+
+In order to make a request as `alice`, we need her access token. Alice's token: `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiYWxpY2UiLCJzdWJvcmRpbmF0ZXMiOltdfQ.Y2EUDNFjnlBKqzr_ix5ELDSOZ_eIYfMeO8W56F9xlkw`.
```

Comment:
For the sake of the tutorial, something like: "To make requests as `alice`, we need her access token:" and then a code block that sets the token into an environment variable:

```bash
export TOKEN=...
```

Then below, in the cURL examples, can refer to $TOKEN, $ALICE_TOKEN, etc.